### DISCUSSION TITLE
A friendly wish for more project metainformation

### DISCUSSION THREAD

**Discussion by hsunner** [commonism/aiopenapi3#136] [General]
I am creating a Python library & CLI for accessing an internal REST interface at my employer. Your library appeared to be the best choice a few month's back when I started and it's been very reassuring to see the heavy development and the rapid, friendly and helpful responses to issues since then. However, as I'm increasingly committing us to depend on your project, it would be valuable to

- have some indication of your plans and goals for this project and possibly the background for it
- know when we (your users in general) should start using the new version (i.e. when it's sufficiently "ready")
- generally understand what expectations and wishes you have on your users and potential contributors

Everything is obviously at your discretion.

**Comment by commonism** [comment]
**Background** - same as you … `I am creating a Python library & CLI for accessing an internal REST interface at my employer.`

I do not like generators, especially when working code-first APIs, so I started with https://github.com/Dorthu/openapi3.
Dorthu/openapi3 has been difficult to work with, I've had some [accepted PRs](https://github.com/Dorthu/openapi3/pulls?q=is%3Apr+author%3Acommonism) with it but the shift (via dataclasses) towards pydantic was not merged https://github.com/Dorthu/openapi3/pull/62 (December 2021), so I choose to fork (January 2022), retried a PR (February 2022) https://github.com/Dorthu/openapi3/pull/69, unlikely.

So one part of my motivation is having a capable python library for Swagger 2.0/OpenAPI 3.0/3.1/3.x, flexible enough to get things done with real world services. Something to work with misbehaving services, invalid/* description documents. The APIs you are presented when integrating something.
The other part is enjoying technology.

**Goals** - best dynamic client library, not to feature creep, keeping things reasonable. 
**Plans** - continue as long as it is relevant as a library/standard wrt. to better alternatives

Lately the pydantic v2 changes required some heavy lifting to get done -the major development activities you referred to-, but well worth it.
A real improvement to a already great library.


**Releases**
I try to be careful with releases, I rely on them myself.
Due to pydantic v2 the changes in 0.3.0 are the most fundamental I can even think of, but 0.3.0 is a great improvement and I tried to make it a drop in replacement.
I intend to have pre-releases sit for a week, no complains, release.
General - if the unit tests do not fail, you should be covered, it not, let's extend the unit tests.

**user-base** - currently small,  easy to work with
The stats from pypi/gh/rtd say it's marginal, e.g. [compared](https://pypistats.org/packages/aiopenapi3) to [openapi3](https://pypistats.org/packages/openapi3) but most of this will be CI jobs somewhere?
This may change if Dorthu/openapi3 does not work well with the FastAPI v0.100.x+ OpenAPI 3.1 requirements, we'll see.

**Expectations/whishes** to users - provide the details required to reproduce issues, motivations for your use-case. 

e.g. I consider the [chunk_iterator](https://github.com/commonism/aiopenapi3/issues/126#issuecomment-1679100612) snippet a use-case worth supporting such as `Request.iterator(data, parameters)`. It's great if you need it, not complex enough to avoid to maintaining and preferable to expose as stable API than require use of internal API to get it done, and chunked_iterator combined with ijson allows parsing large lists item by item to reduce memory footprint.


**potential contributors** - use-case worth covering, pre-commit is fine, code coverage does not decrease, even documentation? welcome.
